feat: Support CRaC and priming of powertools metrics and idempotency-dynamodb

subhash686 opened the pull request:
**Issue #, if available:**
https://github.com/aws-powertools/powertools-lambda-java/issues/1588

## Description of changes:
<!--- One or two sentences as a summary of what's being changed -->
- Added CRaC support in powertools-common, following the guide and blogs mentioned on the ticket.
- Added automatic priming for powertools-metrics by preloading all the required classes, the list of classes is fetched from the JVM at compile time.
- Added manual priming for dynamodb persistence store

**Checklist**

<!--- Leave unchecked if your change doesn't seem to apply --> 

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda-java/#tenets)
* [x] Update tests
* [ ] Update docs
* [x] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Breaking change checklist

<!--- Ignore if it's not a breaking change -->

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

subhash686 commented:
@phipag Looks like v2 is now merged and released, so I resolved conflicts to rebase to `main`. Let me know if you have any other release plans for this.

phipag commented:
Thanks @subhash686 and apologize the delay. I was very busy with the v2 release. 

Your PR is next on my list and we'll try to get it merged for the next minor release 2.1.0.

phipag commented:
Hey @subhash686, I didn't forget you. This week is very busy but I'll do my best to test your implementation very soon. 

In the meantime, can you have a look at the 4 Sonar issued and either resolve them or mark them as not applicable if you think that is the case?

subhash686 commented:
> Hey @subhash686 , I am reviewing your PR now. Can you provide some details how you generated the txt file of the classes loaded? I see you say:
> 
> > the list of classes is fetched from the JVM at compile time.
> 
> Can you provide the compilation command you used for this?
> 
> Ideally, I would like to reproduce the process for how to generate and then use the txt file of loaded classes.

Here are the steps I followed
- Ran following commands to generate the loaded classes in the project root directory
`> export MAVEN_OPTS="-Xlog:class+load=info:classloaded.txt"`
`> mvn install -pl powertools-metrics -am` 

- File has each line in the format `[0.054s][info][class,load] java.lang.Object source: shared objects file`
- Used regex to remove all the prefix extra characters `[\[\]a-z0-9.,]+ ` and a simple replace the text` source: shared objects file` which is at the end of each line.
- Moved the file to powertools-metrics `resources` directory.

phipag commented:
Hey @subhash686,

here is my first update. I tested the classesload.txt (automatic priming) implementation that you contributed and found some issues:
1. This is not running when included in the `LambdaMetricsAspect` that is annotated with `@Aspect`. I moved the priming code for loading the classes into the [EmfMetricsLogger.java](https://github.com/aws-powertools/powertools-lambda-java/blob/8a040ac87a2a31662b7ac8980601b4d821cc29d8/powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/internal/EmfMetricsLogger.java) which worked! I verified this by adding a log statement in the beforeCheckpoint and afterRestore methods and verified in a Snapstart deployment that this gets logged.
2. The classesloaded.txt file does not represent the classes loaded at runtime. Since you generated the file during `mvn install -pl powertools-metrics -am` this will yield the loaded classes during all maven stages part of the install goal. This does not achieve the intended goal which is to pre-load all classes that are essential at runtime of the application. For example, if you search in the file for "MetricsFactory" nothing will show up but this class is mandatory to be loaded at runtime when using the Metrics module.
    - What we do for GraalVM is documented [here](https://github.com/aws-powertools/powertools-lambda-java/blob/main/GraalVM.md). We run the GraalVM tracing agent as part of the unit tests. This way we can actually get the metadata for the loaded classes at runtime (or at least an approximation of it). I envision we can do the same thing here by tracing the loaded classes during execution of the unit tests.  
    - If you checkout @msailes example [here](https://github.com/marksailes/snapstart-priming-guide/blob/main/examples/emf/emf-client-example/src/main/resources/classes.txt) it will have those classes like `MetricsLogger` etc.

For issue (2), I think this needs a completely different approach for generating the classesloaded.txt. We need to implement a mechanism to capture the classes at runtime not the classes used by maven during compilation. Possibly, the GraalVM mechanism can be re-used in some form?

I will now review the Idempotency manual priming that you implemented and will get back to you again. Just wanted to share these preliminary findings. Let me know if you have any suggestion / idea about this. I am happy to join a meeting with you to brainstorm more?

subhash686 commented:
Hi @phipag 
Thanks so much for the time you spent adjusting the code and thorough suggestions. 

For (2), I created a lambda handler with FlushMetrics annotation and used it in a test to generated the classes again. Used GraalVM profile to run all tests with the -Xlog JVM option. I see the runtime classes now with some extra test classes.
I replaced the classesloaded file for now, so we can continue testing. I am happy to meet to brainstorm the approach to generate the class list. 

I will fix (1) once you get a chance to review it all.

phipag commented:
I performed the same load test for the Classloading using the new txt file you provided @subhash686. Here are my results:

### Metrics priming (Classloading)

Tested using example at https://github.com/aws-powertools/powertools-lambda-java/blob/984036ca2e3afb92b2c838bcfa2e48a51c3d7a64/examples/powertools-examples-core-utilities/sam (removed tracer and logger).

We can see ~7% better performance using classloading priming.

#### Without priming

| snapstart_restore_count | avg_restore_ms | min_restore_ms | max_restore_ms | p50_restore_ms | p90_restore_ms | p95_restore_ms | p99_restore_ms |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 400 | 634.9481 | 443.29 | 966.02 | 623.1474 | 730.4813 | 778.7357 | 894.8018 |

#### With priming

| snapstart_restore_count | avg_restore_ms | min_restore_ms | max_restore_ms | p50_restore_ms | p90_restore_ms | p95_restore_ms | p99_restore_ms |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 405 | 586.1477 | 345.14 | 944.56 | 581.0391 | 678.4024 | 718.1742 | 831.0078 |

phipag commented:
Next steps:
- Address issues (1) and (2)
- Brainstorm in a meeting (@phipag and @subhash686) to understand best process for generating classloaded.txt files
- Address PR comments (move Idempotency priming logic to DDB datastore and increase priming scope)
- Merge PR

subhash686 commented:
@phipag It was great meeting with you. 
- I moved the class priming to `MetricsFactory`
- Moved invoke priming to `DynamoDBPersistenceStore` and extended the scope to add and delete record as well.
- Added a new maven profile to generate classes
- Documented it all in  Priming.md

phipag commented:
Awesome, thanks @subhash686. I'll review this by the end of the week.

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+Priming is the process of preloading dependencies and initializing resources during the INIT phase, rather than during the INVOKE phase to further optimize startup performance with SnapStart. 
+This is required because Java frameworks that use dependency injection load classes into memory when these classes are explicitly invoked, which typically happens during Lambda’s INVOKE phase.
+
+This documentation provides guidance for automatic class priming in AWS Lambda Powertools Java modules.
```
```suggestion
This documentation provides guidance for automatic class priming in Powertools for AWS Lambda Java modules.
```

subhash686 replied:
Modified per the suggestion.

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+
+
+## Implementation Steps
+Classes are proactively loaded using Java runtime hooks which are part of the open source CRaC (Coordinated Restore at Checkpoint) project.
```
```suggestion
Classes are proactively loaded using Java runtime hooks which are part of the open source [CRaC (Coordinated Restore at Checkpoint) project](https://openjdk.org/projects/crac/).
```

subhash686 replied:
Modified per the suggestion.

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+
+## Implementation Steps
+Classes are proactively loaded using Java runtime hooks which are part of the open source CRaC (Coordinated Restore at Checkpoint) project.
+This implementation uses this hook, called `beforeCheckpoint()`, to prime SnapStart-enabled Java functions via Class Priming.
```
```suggestion
Implementations across the project use the `beforeCheckpoint()` hook, to prime Snapstart-enabled Java functions via Class Priming.
```

subhash686 replied:
Modified per the suggestion.

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+## Implementation Steps
+Classes are proactively loaded using Java runtime hooks which are part of the open source CRaC (Coordinated Restore at Checkpoint) project.
+This implementation uses this hook, called `beforeCheckpoint()`, to prime SnapStart-enabled Java functions via Class Priming.
+In order to generate the `classloaded` files for Powertools java module, follow these general steps.
```
```suggestion
In order to generate the `classesloaded.txt` file for a Java module in this project, follow these general steps.
```

subhash686 replied:
Modified per the suggestion.

phipag commented on Priming.md:
````diff
@@ -0,0 +1,58 @@
+   ```shell
+      -Xlog:class+load=info:classesloaded.txt
+      ```
+    - You can find an example of this in profile `generate-classesloaded-file` in this [pom.xml](powertools-metrics/pom.xml).
````
```suggestion
    - You can find an example of this `generate-classesloaded-file` profile in this [pom.xml](powertools-metrics/pom.xml).
```

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+3. **Cleanup the file**
+    - The classes loaded file generated in Step 2 has the format
+   `[0.054s][info][class,load] java.lang.Object source: shared objects file` 
+   but we are only interested in `java.lang.Object` the fully qualified class name.
```
```suggestion
   but we are only interested in `java.lang.Object` – the fully qualified class name.
```

subhash686 replied:
Modified per the suggestion.

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+      - `^\[[\[\]0-9.a-z,]+ ` (to replace the left part)
+      - `( source: )[0-9a-z :/._$-]+` (to replace the right part)
+
+4. **Place the file**
```
```suggestion
4. **Add file to resources**
```

subhash686 replied:
Modified per the suggestion.

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+      - `( source: )[0-9a-z :/._$-]+` (to replace the right part)
+
+4. **Place the file**
+    -  Move the cleanedup file containing to the corresponding resources directory of the module. See [example](powertools-metrics/src/resources/classesloaded.txt).
```
```suggestion
    -  Move the cleaned-up file to the corresponding `src/main/resources` directory of the module. See [example](powertools-metrics/src/main/resources/classesloaded.txt).
```

subhash686 replied:
Modified per the suggestion.

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+4. **Place the file**
+    -  Move the cleanedup file containing to the corresponding resources directory of the module. See [example](powertools-metrics/src/resources/classesloaded.txt).
+
+5. **Register and checkpoint**
```
Let's clarify this section by stating the following information:
- Mention that ClassPreLoader is implemented in `powertools-common` module
- Mention that this cannot be done in an AspectJ aspect
- Mention that this needs to be done in a suitable class. A class is suitable if it is the "entrypoint" for a module – i.e. `MetricsFactory` for Metrics module.

subhash686 replied:
Added language to clarify entrypoint selection to register with CRaC.

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+5. **Register and checkpoint**
+    - Register the CRaC Resource in the constructor of the main class.
+    - Add the `beforeCheckpoint()` hook in the same class to invoke `ClassPreLoader.preloadClasses()`.
+    - This will ensure that the classes are loaded during the INIT phase of the Lambda function. See [example](powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsFactory.java)
```
```suggestion
    - This will ensure that the classes are already pre-loaded by the Snapstart RESTORE operation leading to a shorter INIT duration. See [example](powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsFactory.java)
```

phipag commented on Priming.md:
```diff
@@ -0,0 +1,58 @@
+
+## Known Issues
+- This is a manual process at the moment, but it can be automated in the future.
+- Because the file is generated while running tests, it includes test classes as well.
```
Let's explain here why this is not a problem. It is not a problem because all classes that are not found are ignored by the ClassPreLoader and the `beforeCheckpoint()` hook is not time sensitive since it only runs once when a new Lambda function version gets published.

subhash686 replied:
Modified per the suggestion.

phipag reviewed (changes requested):
Thanks @subhash686. I added some cosmetic comments in the Priming documentation. I am deploying your changes to a Snapstart Lambda for testing now. I will get back with the results later.

subhash686 commented on powertools-common/src/main/java/software/amazon/lambda/powertools/common/internal/ClassPreLoader.java:
```diff
@@ -0,0 +1,77 @@
+                URL url = files.nextElement();
+                URLConnection conn = url.openConnection();
+                conn.setUseCaches(false);
+                InputStream is = conn.getInputStream();
```
I have the try-with-resources statement in the method below which ensures proper closing. Do you see any concern with that approach?

phipag replied:
The input stream in the method below is a second input stream which is closed properly but I think this one needs proper closing as well. 

```java
try(InputStream is = conn.getInputStream();) {
                preloadClassesFromStream(is);
}
```

subhash686 replied:
Because InputStream is passed by reference and the try-with-resource handles the resource variables that are previously initialized as well, so the InputStream from line 45 must get closed properly. 

```java
private static void preloadClassesFromStream(InputStream is) {
        try (is; 
```

If you think its confusing doing it this way, I will remove `is` from this try block and move it up to to the caller. I dont think its right to include it twice.

phipag replied:
You are right, I see the `is;` part of the `try` now. Let's keep it as is then. I will mark the Sonar finding as false positive.

subhash686 commented on powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsFactory.java:
```diff
@@ -68,4 +72,14 @@ public static synchronized void setMetricsProvider(MetricsProvider metricsProvid
+
+    @Override
+    public void beforeCheckpoint(org.crac.Context<? extends Resource> context) throws Exception {
+        ClassPreLoader.preloadClasses();
```
Great idea. Added it.

subhash686 commented on powertools-common/src/main/java/software/amazon/lambda/powertools/common/internal/ClassPreLoader.java:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Used to preload classes to support automatic priming for SnapStart
+ */
+public class ClassPreLoader {
```
Modified per the suggestion.

subhash686 commented on powertools-idempotency/powertools-idempotency-dynamodb/src/main/java/software/amazon/lambda/powertools/idempotency/persistence/dynamodb/DynamoDBPersistenceStore.java:
```diff
@@ -52,7 +55,7 @@
  * Use the {@link Builder} to create a new instance.
  */
-public class DynamoDBPersistenceStore extends BasePersistenceStore implements PersistenceStore {
+public class DynamoDBPersistenceStore extends BasePersistenceStore implements PersistenceStore, Resource {
```
Modified per the suggestion.

subhash686 commented on powertools-common/src/test/java/software/amazon/lambda/powertools/common/internal/ClassPreLoaderTest.java:
```diff
@@ -0,0 +1,14 @@
+class ClassPreLoaderTest {
+
+    @Test
+    void preloadClasses_shouldIgnoreInvalidClassesAndLoadValidClasses() {
+        // Verify that the missing class did not throw any exception
+        assertDoesNotThrow(ClassPreLoader::preloadClasses);
+    }
```
I will look into this. Did not change anything yet.

phipag replied:
Awesome, thank you!

subhash686 replied:
I have a `classessloaded.txt` file already for tests but I did not test whether valid classess are loaded specially when the file is a mixed bag of valid and invalid classes. Again thanks for catching that bug. 

Its a little tricky to test the static methods here but I got a little creative with `volatile` to not complicate tests with modifying thread context and cleanup. 

Finally, I checked that my new assertion fails with old code where if one class is not found the rest of the classes are skipped. And it works after fixing that code.

Let me know know if you see any more opportunities here.

subhash686 commented:
@phipag Thanks for your thorough review as always. I pushed the updates. 
I checked the 2 sonar issues and am inclined to accept those. Please let me know if you think otherwise.

phipag commented on powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/MetricsFactory.java:
```diff
@@ -23,11 +26,12 @@
     private static Metrics metrics;
 
     private MetricsFactory() {
+        Core.getGlobalContext().register(this);
```
Yes, correct. The Builder in the Dynamo persistence store instantiates the class itself which is why the priming gets executed. The MetricsFactory however instantiates the provider which is a separate class.

phipag commented on powertools-common/src/main/java/software/amazon/lambda/powertools/common/internal/ClassPreLoader.java:
```diff
@@ -0,0 +1,77 @@
+                final String className = line.stripTrailing();
+                if (!className.isBlank()) {
+                    Class.forName(className, true, ClassPreLoader.class.getClassLoader());
+                }
```
Nice, thanks.

phipag commented:
> @phipag Thanks for your thorough review as always. I pushed the updates. I checked the 2 sonar issues and am inclined to accept those. Please let me know if you think otherwise.

Thanks very much for addressing my comments so quickly @subhash686. I resolved all comments which are fixed now. Please review the open comments.

Regarding Sonarcube, I accepted one issue. Let's address the second one in the next commit. We can extract the class loading into a method like `loadClassIfAvailable(className)`.

subhash686 commented:
> > @phipag Thanks for your thorough review as always. I pushed the updates. I checked the 2 sonar issues and am inclined to accept those. Please let me know if you think otherwise.
> 
> Thanks very much for addressing my comments so quickly @subhash686. I resolved all comments which are fixed now. Please review the open comments.
> 
> Regarding Sonarcube, I accepted one issue. Let's address the second one in the next commit. We can extract the class loading into a method like `loadClassIfAvailable(className)`.
> 
> I tested the code again after your updates in my AWS account. All Snapstart hooks run as expected now without errors.

Fixed this Sonarqube finding.

phipag commented on powertools-common/src/test/java/software/amazon/lambda/powertools/common/internal/ClassPreLoaderTest.java:
```diff
@@ -0,0 +1,34 @@
+class ClassPreLoaderTest {
+
+    // Making this volatile so the Thread Context doesn't need any special handling
+    static volatile boolean dummyClassLoaded = false;
```
Out of interest: Can you explain what happens if don't put volatile here? I do not 100% follow the comment.

subhash686 replied:
I will not be able to modify the dummyClassLoaded as if its a local variable. I am setting it to false and then checking ClassPreLoader::preLoadClasses turns it to true if it loaded successfully. 
I did not witness in handful runs but theoretically there is a possibility of flakiness in tests as the variable change from classloader may not available immediately. 
Alternate option is to set and reset `Thread.currentThread().setContextClassLoader`.

At least thats my understanding, I ran the tests almost 7 times and it worked every time with or without volatile.

phipag replied:
I think I understand now:
1. Test initializes -> `dummyClassLoaded = false` is declared and initialized as `false`
2. ClassPreLoader runs
3. ClassPreLoader loads `DummyClass`
4. Static block triggers and sets `dummyClassLoaded = true` on the `ClassPreLoaderTest` class

It needs to be volatile such that the junit thread can immediatly assert the updated value based on the classloading of `DummyClass` executing the static block. 

Let me know if this makes sense.

subhash686 replied:
You've rightly put it!

subhash686 commented on powertools-logging/powertools-logging-logback/src/test/java/software/amazon/lambda/powertools/logging/internal/LambdaJsonEncoderTest.java:
```diff
@@ -368,7 +368,7 @@ void shouldLogThreadInfo() {
 
         // THEN
-        assertThat(result).contains("\"thread\":\"main\",\"thread_id\":1,\"thread_priority\":5");
+        assertThat(result).contains("\"thread\":\"main\",\"thread_id\":"+ Thread.currentThread().getId() +",\"thread_priority\":5");
```
Test is failing locally as my thread id is 3 but the test is expecting 1.

phipag replied:
Cool. Thanks for catching this!

phipag reviewed (approved):
@subhash686, I think this PR is ready to be merged now. I would like to congratulate you on this major contribution. This is laying the foundation for adding different CRaC priming techniques across this project. Also many thanks for documenting this process in the `Priming.md` doc. 🚀

In the following days, I will create follow-up issues on the items we discussed:
- Sub-issues for each Powertools utility referencing your Priming documentation
- An issue for automating the maintenance of the `classesloaded.txt` file through a GitHub workflow (implementation details to be defined).
